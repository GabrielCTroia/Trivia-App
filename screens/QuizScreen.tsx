import React from 'react';
import {
  StyleSheet,
  View,
} from 'react-native';
import { Constants } from 'expo';

import Colors from '../constants/Colors';
import CommonStyle from '../styles/CommonStyle';
import { StyledButton } from '../components/buttons/StyledButton';
import { Title } from '../components/text/Title';
import { Paragraph } from '../components/text/Paragraph';
import { NavigationScreenProp } from 'react-navigation';

// enum QuestionDiffuculty {
//   easy,
//   hard,
// }

// export class Question {
//   public title: string;
//   public difficulty: QuestionDiffuculty;
//   public correctAnswer: boolean;

//   constructor({}) {

//   }
// }

// export class Answer {
//   public isCorrect: boolean;

//   constructor(public question: Question, answered: boolean) {
//     this.isCorrect = question.
//   }
// }

export interface QuizScreenParams {
  category: string;
  topic: string;

  questions: {
    title: string;
    correctAnswer: boolean;
  }[],
}

export interface QuizScreenProps {
  navigation: NavigationScreenProp<any, QuizScreenParams>;
}



// const passedProps: QuizScreenProps = {
//   category: 'Entertainment',
//   topic: 'Video Games',

//   questions: [
//     {
//       title: 'Is that you Mario?',
//       correctAnswer: true,
//     }
//   ]
// }


export class QuizScreen extends React.Component<QuizScreenProps> {
  static navigationOptions = {
    header: null,
  };

  render() {
    // const x = this.props.navigation.state
    console.log('navigation', this.props.navigation);

    return (
      <View style={styles.container}>
        <Title>Welcome to Trivia Challenge!</Title>
        <Paragraph style={CommonStyle.centerText}>
          You will be presented with 10 True or False questions.
        </Paragraph>
        <Paragraph style={CommonStyle.centerText}>
          Can you score 100%?
        </Paragraph>
        <StyledButton
          title="Begin"
          onPress={(s) => { console.log(s) }}
        />
      </View>
    );
  }

}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.noticeBackground,
    alignContent: "center",
    justifyContent: 'space-between',
    paddingTop: Constants.statusBarHeight,
  },
  titleText: {
    textAlign: "center",
    fontSize: 25,
    fontWeight: 'bold',
  },
});
