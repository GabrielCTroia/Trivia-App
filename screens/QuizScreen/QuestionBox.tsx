import * as React from 'react';
import { StyleSheet, View, Text } from 'react-native';
import { Question } from '../../Api/Questions';
import { screenHorizontalPadding } from '../../styles/Layout';
import * as Typography from '../../styles/Typography';
import * as Colors from '../../styles/Colors';


export type QuestionBoxProps = {
  index: number,
  question: Question,
};

export const QuestionBox: React.FunctionComponent<QuestionBoxProps> = (props) => {
  return (
    <View style={styles.container} >
      <Text style={styles.tipText}>{props.index}</Text>
      <Text style={styles.text}>{props.question.title}</Text>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.whiteColor,
    borderRadius: 26,
    padding: screenHorizontalPadding,
    flex: 1,
    textAlign: 'center',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
  },
  tipText: {
    ...Typography.baseText,
    fontSize: 45,
    fontFamily: 'roboto-black',
    color: Colors.warningColor,
    opacity: .5,
  },
  text: {
    ...Typography.baseText,
    color: Colors.baseTextColor,
    flex: 1,
    paddingLeft: 10,
  }
});
