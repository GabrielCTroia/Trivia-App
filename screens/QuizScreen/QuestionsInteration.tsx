import React, { FunctionComponent } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import { QuestionBox } from './QuestionBox';
import { CardStack } from '../../components/SwipeableCards/CardStack';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import * as Typography from '../../styles/Typography';
import * as Colors from '../../styles/Colors';
import { Question } from '../../Api/Questions';

export type QuestionsInteraction = {
  questions: Question[];
  onAnswer: (id: string, options: boolean) => void;
  totalQuestions: number,
};

export const QuestionsInteraction: FunctionComponent<QuestionsInteraction> = (props) => {
  return (
    <View style={styles.container}>
      <View style={styles.tutorial}>
        <View style={styles.tutorialTextContainer}>
          <Text style={styles.tutorialText}>FALSE</Text>
        </View>
        <View style={styles.swipeTipContainer}>
          <MaterialCommunityIcons
            style={styles.swipeTip}
            name="gesture-swipe-horizontal"
            color={Colors.lightShade}
          />
        </View>
        <View style={styles.tutorialTextContainer}>
          <Text style={styles.tutorialText}>TRUE</Text>
        </View>
      </View>
      <CardStack
        items={props.questions}
        keyExtractor={(item) => item.id}
        renderItem={(item, index) => <QuestionBox question={item} index={index} />}
        onSwipeLeft={(id) => props.onAnswer(id, false)}
        onSwipeRight={(id) => props.onAnswer(id, true)}
        totalItemsCount={props.totalQuestions}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
  },
  tutorial: {
    flex: .4,
    flexDirection: 'row',
    justifyContent: 'center',
  },
  swipeTipContainer: {
    justifyContent: 'center',
  },
  swipeTip: {
    fontSize: 80,
  },
  tutorialTextContainer: {
    justifyContent: 'center',
  },
  tutorialText: {
    ...Typography.baseText,
    color: Colors.baseTextColor,
  },
});
