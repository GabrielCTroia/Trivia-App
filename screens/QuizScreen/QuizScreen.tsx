import React from 'react';
import R from 'ramda';
import { StyleSheet, View, Text, StatusBar, } from 'react-native';
import { NavigationScreenProp } from 'react-navigation';
import { Question } from '../../Api/Questions';
import { StyledButton, ButtonTypes } from '../../components/buttons/StyledButton';
import { AnsweredQuestion } from '../ResultsScreen/AnsweredQuestionItem';
import { HeaderBar } from '../../components/HeaderBar/HeaderBar';
import * as Colors from '../../styles/Colors';
import * as Layout from '../../styles/Layout';
import * as Typography from '../../styles/Typography';
import { QuestionsInteraction } from './QuestionsInteration';


export interface QuizScreenParams {
  category: string;
  questions: Question[],
}

export interface QuizScreenProps {
  navigation: NavigationScreenProp<any, QuizScreenParams>;
}

interface State {
  givenAnswersById: { [k: string]: boolean };
  quizEnded: boolean;
}

// TODO: For some reason the QuizScreenParams are not used/inferred from the react-navigation lib. Investigate further!
// Temporary solution: Use a function that specializes in extracting and annotating the questions.
const getQuestionsFromNavParams = (props: QuizScreenProps): Question[] => {
  return props.navigation.state.params.questions || [];
}

export class QuizScreen extends React.Component<QuizScreenProps, State> {
  static navigationOptions = {
    header: null,
  };

  constructor(props: QuizScreenProps) {
    super(props);

    this.state = {
      givenAnswersById: {},
      quizEnded: false,
    }
  }

  private answer(questionId: string, option: boolean) {


    const nextAnsweresById = R.merge(this.state.givenAnswersById, {
      [questionId]: option,
    });

    this.setState({
      givenAnswersById: nextAnsweresById,
      quizEnded: R.keys(nextAnsweresById).length === getQuestionsFromNavParams(this.props).length,
    });
  }

  private getAnsweredQuestions(): AnsweredQuestion[] {
    return getQuestionsFromNavParams(this.props)
      .map((q) => ({ ...q, givenAnswer: this.state.givenAnswersById[q.id] }))
  }

  private renderQuestions() {
    if (this.state.quizEnded) {
      return null;
    }

    const questions = getQuestionsFromNavParams(this.props);

    // Only render the unanswered questions.
    // In the future the QuestionInterction (CardStack) could do it's own optimization
    //  but for now this is enough. 
    const onlyUnanswered = questions.filter((q) => this.state.givenAnswersById[q.id] === undefined);

    return (
      <QuestionsInteraction
        questions={onlyUnanswered}
        onAnswer={(id, opt) => this.answer(id, opt)}
        totalQuestions={questions.length}
      />
    );
  }

  private renderResultsButton() {
    if (!this.state.quizEnded) {
      return null;
    }

    return <StyledButton
      title="See Results"
      buttonType={ButtonTypes.warning}
      onPress={() => this.props.navigation.navigate('Results', {
        answeredQuestions: this.getAnsweredQuestions(),
      })}
    />
  }

  render() {
    const { category = '' } = this.props.navigation.state.params || {};

    const [topic, subtopic] = category.split(':');

    return (
      <View style={styles.container}>
        <StatusBar barStyle="dark-content" />
        <View style={styles.screen}>
          <View>
            <HeaderBar />
            <Text style={styles.header}>{topic}</Text>
            <Text style={styles.subheader}>{subtopic}</Text>
          </View>
          <View style={styles.main}>
            <View style={styles.content}>
              {this.renderQuestions()}
              {this.renderResultsButton()}
            </View>
          </View>
        </View>
      </View>
    );
  }

}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.backgroundColor,
  },
  screen: {
    ...Layout.screenLayout,
    flex: .87,

    backgroundColor: Colors.foregroundColor,
    borderBottomLeftRadius: Layout.screenBorderRadius,
    borderBottomRightRadius: Layout.screenBorderRadius,
  },
  main: {
    flex: 1,
    justifyContent: 'center',
  },
  header: {
    ...Typography.headerText,
  },
  subheader: {
    ...Typography.subheaderText,
  },
  content: {
    height: Layout.screenHeight / 2,
    justifyContent: 'space-around',
  },
});
