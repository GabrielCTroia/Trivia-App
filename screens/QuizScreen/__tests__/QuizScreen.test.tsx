import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { QuizScreen } from '../QuizScreen';
import { NavigationScreenProp } from 'react-navigation';

it('renders correctly', () => {
  const mockNavigation = {
    state: {
      params: {},
    }
  } as unknown as NavigationScreenProp<any, any>;

  const tree = renderer.create(<QuizScreen navigation={mockNavigation} />).toJSON();

  expect(tree).toMatchSnapshot();
});
