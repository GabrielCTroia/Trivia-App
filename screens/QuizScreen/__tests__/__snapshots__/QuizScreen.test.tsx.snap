// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#4044C9",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderBottomLeftRadius": 45,
        "borderBottomRightRadius": 45,
        "flex": 0.87,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 20,
      }
    }
  >
    <View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
            "paddingBottom": 10,
            "paddingTop": 20,
          }
        }
      >
        <Text />
      </View>
      <Text
        style={
          Object {
            "alignItems": "center",
            "color": "#F2994A",
            "display": "flex",
            "flexDirection": "row",
            "fontFamily": "roboto-black",
            "fontSize": 40,
            "justifyContent": "center",
          }
        }
      >
        
      </Text>
      <Text
        style={
          Object {
            "alignItems": "center",
            "color": "#4044C9",
            "display": "flex",
            "flexDirection": "row",
            "fontFamily": "roboto-regular",
            "fontSize": 32,
            "justifyContent": "center",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "height": 667,
            "justifyContent": "space-around",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <View
            style={
              Object {
                "flex": 0.4,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignItems": "center",
                    "color": "#828282",
                    "display": "flex",
                    "flexDirection": "row",
                    "fontFamily": "roboto-regular",
                    "fontSize": 16,
                    "justifyContent": "center",
                  }
                }
              >
                FALSE
              </Text>
            </View>
            <View
              style={
                Object {
                  "justifyContent": "center",
                }
              }
            >
              <Text />
            </View>
            <View
              style={
                Object {
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignItems": "center",
                    "color": "#828282",
                    "display": "flex",
                    "flexDirection": "row",
                    "fontFamily": "roboto-regular",
                    "fontSize": 16,
                    "justifyContent": "center",
                  }
                }
              >
                TRUE
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignContent": "center",
                "flex": 1,
                "justifyContent": "flex-start",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "flex": 1,
                    "justifyContent": "flex-end",
                  },
                  undefined,
                ]
              }
            />
            <View
              style={
                Array [
                  Object {
                    "alignContent": "center",
                    "height": 15,
                    "justifyContent": "center",
                  },
                  Object {
                    "justifyContent": "flex-end",
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignSelf": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
